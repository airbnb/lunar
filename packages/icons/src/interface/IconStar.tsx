import React from 'react';
import withIcon, { Props } from '../withIcon';

function IconStar(props: Props) {
  return (
    <svg viewBox="0 0 24 24" {...props}>
      <path d="M19.649 9.295l-4.84-.42-1.89-4.45c-.34-.81-1.5-.81-1.84 0l-1.89 4.46-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.73 3.67-3.18c.67-.58.32-1.68-.56-1.75zm-7.65 6.36l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38 1.7-4.03 1.71 4.04 4.38.38-3.32 2.88 1 4.28z" />
    </svg>
  );
}

export default withIcon('IconStar')(IconStar);
