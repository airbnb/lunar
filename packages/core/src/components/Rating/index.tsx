import React from 'react';
import { between } from 'airbnb-prop-types';
import IconStar from '@airbnb/lunar-icons/lib/interface/IconStar';
import IconStarFull from '@airbnb/lunar-icons/lib/interface/IconStarFull';
import IconStarHalf from '@airbnb/lunar-icons/lib/interface/IconStarHalf';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import Spacing from '../Spacing';
import Text from '../Text';

const NUMBER_OF_STARS = 5;
const STARS = new Array(NUMBER_OF_STARS).fill(null);

export type Props = {
  /** Increase font size to large. */
  large?: boolean;
  /** Decrease font size to small. */
  micro?: boolean;
  /** Star rating number. */
  rating?: number;
  /** Review display number. */
  reviews?: string | number;
  /** Decrease font size to small. */
  small?: boolean;
};

/** Display a star rating and review count. */
export class Rating extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    rating: between({ gte: 0, lte: NUMBER_OF_STARS }),
  };

  static defaultProps = {
    large: false,
    micro: false,
    rating: 0,
    reviews: '',
    small: false,
  };

  render() {
    const { large, micro, rating = 0, reviews, small, styles, theme } = this.props;

    return (
      <Text large={large} micro={micro} small={small}>
        <div
          {...css(
            styles.ratingContainer,
            large && styles.ratingContainer_large,
            micro && styles.ratingContainer_micro,
            small && styles.ratingContainer_small,
          )}
        >
          {STARS.map((star, index) => {
            const key = `star-${index}`;
            const color =
              rating > index || (rating > index && rating < index + 1)
                ? theme!.color.core.primary[3]
                : theme!.color.core.neutral[3];
            let Star = IconStar;

            if (rating > index && rating < index + 1) {
              Star = IconStarHalf;
            } else if (rating > index) {
              Star = IconStarFull;
            }

            return <Star key={key} color={color} size="1.15em" decorative />;
          })}

          {reviews && <Spacing left={reviews ? 0.5 : 0}>{reviews}</Spacing>}
        </div>
      </Text>
    );
  }
}

export default withStyles(
  ({ font }) => ({
    ratingContainer: {
      display: 'flex',
      alignItems: 'center',
      height: font.textRegular.lineHeight,
    },

    ratingContainer_large: {
      height: font.textLarge.lineHeight,
    },

    ratingContainer_micro: {
      height: font.textMicro.lineHeight,
    },

    ratingContainer_small: {
      height: font.textSmall.lineHeight,
    },
  }),
  {
    passThemeProp: true,
  },
)(Rating);
