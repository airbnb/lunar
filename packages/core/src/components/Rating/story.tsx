import React from 'react';
import Rating from '.';

export default {
  title: 'Core/Rating',
  parameters: {
    inspectComponents: [Rating],
  },
};

export function starRatings() {
  return (
    <>
      <Rating rating={0} />
      <Rating rating={0.5} />
      <Rating rating={1} />
      <Rating rating={1.5} />
      <Rating rating={2} />
      <Rating rating={2.5} />
      <Rating rating={3} />
      <Rating rating={3.5} />
      <Rating rating={4} />
      <Rating rating={4.5} />
      <Rating rating={5} />
    </>
  );
}

starRatings.story = {
  name: 'Star ratings.',
};

export function withReviews() {
  return <Rating rating={4.5} reviews={100} />;
}

withReviews.story = {
  name: 'With reviews.',
};

export function withDifferentSizing() {
  return (
    <>
      <Rating micro rating={3.5} reviews={100} />
      <Rating small rating={3.5} reviews={100} />
      <Rating rating={3.5} reviews={100} />
      <Rating large rating={3.5} reviews={100} />
    </>
  );
}

withDifferentSizing.story = {
  name: 'With different sizing.',
};
