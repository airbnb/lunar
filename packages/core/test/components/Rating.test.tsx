import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import IconStar from '@airbnb/lunar-icons/lib/interface/IconStar';
import IconStarFull from '@airbnb/lunar-icons/lib/interface/IconStarFull';
import IconStarHalf from '@airbnb/lunar-icons/lib/interface/IconStarHalf';
import Rating from '../../src/components/Rating';

describe('<Rating />', () => {
  it('renders default', () => {
    const wrapper = shallowWithStyles(<Rating rating={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders micro', () => {
    const wrapper = shallowWithStyles(<Rating micro rating={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders small', () => {
    const wrapper = shallowWithStyles(<Rating small rating={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(<Rating large rating={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders reviews', () => {
    const wrapper = shallowWithStyles(<Rating reviews="100" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders all empty stars for a rating of 0', () => {
    const wrapper = shallowWithStyles(<Rating rating={0} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(0);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(5);
  });

  it('renders expected stars for a rating between 0 and 1', () => {
    const wrapper = shallowWithStyles(<Rating rating={0.5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(0);
    expect(wrapper.find(IconStarHalf)).toHaveLength(1);
    expect(wrapper.find(IconStar)).toHaveLength(4);
  });

  it('renders expected stars for a rating of 1', () => {
    const wrapper = shallowWithStyles(<Rating rating={1} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(1);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(4);
  });

  it('renders expected stars for a rating between 1 and 2', () => {
    const wrapper = shallowWithStyles(<Rating rating={1.5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(1);
    expect(wrapper.find(IconStarHalf)).toHaveLength(1);
    expect(wrapper.find(IconStar)).toHaveLength(3);
  });

  it('renders expected stars for a rating of 2', () => {
    const wrapper = shallowWithStyles(<Rating rating={2} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(2);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(3);
  });

  it('renders expected stars for a rating between 2 and 3', () => {
    const wrapper = shallowWithStyles(<Rating rating={2.5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(2);
    expect(wrapper.find(IconStarHalf)).toHaveLength(1);
    expect(wrapper.find(IconStar)).toHaveLength(2);
  });

  it('renders expected stars for a rating of 3', () => {
    const wrapper = shallowWithStyles(<Rating rating={3} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(3);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(2);
  });

  it('renders expected stars for a rating between 3 and 4', () => {
    const wrapper = shallowWithStyles(<Rating rating={3.5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(3);
    expect(wrapper.find(IconStarHalf)).toHaveLength(1);
    expect(wrapper.find(IconStar)).toHaveLength(1);
  });

  it('renders expected stars for a rating of 4', () => {
    const wrapper = shallowWithStyles(<Rating rating={4} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(4);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(1);
  });

  it('renders expected stars for a rating between 4 and 5', () => {
    const wrapper = shallowWithStyles(<Rating rating={4.5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(4);
    expect(wrapper.find(IconStarHalf)).toHaveLength(1);
    expect(wrapper.find(IconStar)).toHaveLength(0);
  });

  it('renders expected stars for a rating of 5', () => {
    const wrapper = shallowWithStyles(<Rating rating={5} />);

    expect(wrapper.find(IconStarFull)).toHaveLength(5);
    expect(wrapper.find(IconStarHalf)).toHaveLength(0);
    expect(wrapper.find(IconStar)).toHaveLength(0);
  });
});
