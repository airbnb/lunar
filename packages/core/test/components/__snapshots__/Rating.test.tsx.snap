// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rating /> renders default 1`] = `
<Text
  large={false}
  micro={false}
  small={false}
>
  <div
    className="ratingContainer_au0dw8"
  >
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-0"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-1"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-2"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-3"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-4"
      size="1.15em"
    />
  </div>
</Text>
`;

exports[`<Rating /> renders large 1`] = `
<Text
  large={true}
  micro={false}
  small={false}
>
  <div
    className="ratingContainer_au0dw8-o_O-ratingContainer_large_1al5d6z"
  >
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-0"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-1"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-2"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-3"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-4"
      size="1.15em"
    />
  </div>
</Text>
`;

exports[`<Rating /> renders micro 1`] = `
<Text
  large={false}
  micro={true}
  small={false}
>
  <div
    className="ratingContainer_au0dw8-o_O-ratingContainer_micro_l5dkxe"
  >
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-0"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-1"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-2"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-3"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-4"
      size="1.15em"
    />
  </div>
</Text>
`;

exports[`<Rating /> renders reviews 1`] = `
<Text
  large={false}
  micro={false}
  small={false}
>
  <div
    className="ratingContainer_au0dw8"
  >
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-0"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-1"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-2"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-3"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-4"
      size="1.15em"
    />
    <Spacing
      left={0.5}
    >
      100
    </Spacing>
  </div>
</Text>
`;

exports[`<Rating /> renders small 1`] = `
<Text
  large={false}
  micro={false}
  small={true}
>
  <div
    className="ratingContainer_au0dw8-o_O-ratingContainer_small_9f436v"
  >
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-0"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-1"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-2"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-3"
      size="1.15em"
    />
    <IconStar
      color="#C2C2C2"
      decorative={true}
      key="star-4"
      size="1.15em"
    />
  </div>
</Text>
`;
